ChatUI/GridCtrl_src/InPlaceEdit.cpp

```diff
@@ -0,0 +1,263 @@
+//         10 Apr 1999  Enter, Tab and Esc key prob fixed (Koay Kah Hoe)
+//                      Workaround for bizzare "shrinking window" problem in CE
+//
+/////////////////////////////////////////////////////////////////////////////
+
+#include "stdafx.h"
+#include "TCHAR.h"
+#include "InPlaceEdit.h"
+
+#include "GridCtrl.h"
+
+#ifdef _DEBUG
+#define new DEBUG_NEW
+#undef THIS_FILE
+static char THIS_FILE[] = __FILE__;
+#endif
+
+/////////////////////////////////////////////////////////////////////////////
+// CInPlaceEdit
+
+CInPlaceEdit::CInPlaceEdit(CWnd* pParent, CRect& rect, DWORD dwStyle, UINT nID,
+                           int nRow, int nColumn, CString sInitText, 
+                           UINT nFirstChar)
+{
+    m_sInitText     = sInitText;
+    m_nRow          = nRow;
+    m_nColumn       = nColumn;
+    m_nLastChar     = 0; 
+    m_bExitOnArrows = (nFirstChar != VK_LBUTTON);    // If mouse click brought us here,
+                                                     // then no exit on arrows
+
+    m_Rect = rect;  // For bizarre CE bug.
+    
+    DWORD dwEditStyle = WS_BORDER|WS_CHILD|WS_VISIBLE| ES_AUTOHSCROLL //|ES_MULTILINE
+        | dwStyle;
+    if (!Create(dwEditStyle, rect, pParent, nID)) return;
+    
+    SetFont(pParent->GetFont());
+    
+    SetWindowText(sInitText);
+    SetFocus();
+    
+    switch (nFirstChar){
+        case VK_LBUTTON: 
+        case VK_RETURN:   SetSel((int)_tcslen(m_sInitText), -1); return;
+        case VK_BACK:     SetSel((int)_tcslen(m_sInitText), -1); break;
+        case VK_TAB:
+        case VK_DOWN: 
+        case VK_UP:   
+        case VK_RIGHT:
+        case VK_LEFT:  
+        case VK_NEXT:  
+        case VK_PRIOR: 
+        case VK_HOME:
+        case VK_SPACE:
+        case VK_END:      SetSel(0,-1); return;
+        default:          SetSel(0,-1);
+    }
+
+    // Added by KiteFly. When entering DBCS chars into cells the first char was being lost
+    // SenMessage changed to PostMessage (John Lagerquist)
+    if( nFirstChar < 0x80)
+        PostMessage(WM_CHAR, nFirstChar);   
+    else
+        PostMessage(WM_IME_CHAR, nFirstChar);
+}
+
+CInPlaceEdit::~CInPlaceEdit()
+{
+}
+
+BEGIN_MESSAGE_MAP(CInPlaceEdit, CEdit)
+    //{{AFX_MSG_MAP(CInPlaceEdit)
+    ON_WM_KILLFOCUS()
+    ON_WM_CHAR()
+    ON_WM_KEYDOWN()
+    ON_WM_GETDLGCODE()
+    ON_WM_CREATE()
+    //}}AFX_MSG_MAP
+END_MESSAGE_MAP()
+
+////////////////////////////////////////////////////////////////////////////
+// CInPlaceEdit message handlers
+
+// If an arrow key (or associated) is pressed, then exit if
+//  a) The Ctrl key was down, or
+//  b) m_bExitOnArrows == TRUE
+void CInPlaceEdit::OnKeyDown(UINT nChar, UINT nRepCnt, UINT nFlags) 
+{
+    if ((nChar == VK_PRIOR || nChar == VK_NEXT ||
+        nChar == VK_DOWN  || nChar == VK_UP   ||
+        nChar == VK_RIGHT || nChar == VK_LEFT) &&
+        (m_bExitOnArrows || GetKeyState(VK_CONTROL) < 0))
```

Comment:
too big if-case. Maybe, it would be better to move that check into some boolean function?